File: android/EyeTimeTrackerAndroid/src/com/eyetimetracker/android/LegacyUsageSegments.java
package com.eyetimetracker.android;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class LegacyUsageSegments {
    public static final String SOURCE = "legacy-summary";

    private LegacyUsageSegments() {
    }

    public static List<UsageSegment> normalizeEffectiveSegments(
            List<UsageSegment> storedSegments,
            List<DailySummary> localSummaries,
            String deviceId,
            String platform) {
        List<UsageSegment> stored = storedSegments == null ? new ArrayList<>() : storedSegments;
        List<UsageSegment> modern = new ArrayList<>();
        List<UsageSegment> peerLegacy = new ArrayList<>();
        for (UsageSegment segment : stored) {
            if (segment == null) {
                continue;
            }
            if (SOURCE.equals(segment.source)) {
                if (!safe(deviceId).equals(segment.deviceId)) {
                    peerLegacy.add(segment);
                }
            } else {
                modern.add(cloneSegment(segment));
            }
        }

        List<UsageSegment> effective = new ArrayList<>();
        effective.addAll(modern);
        effective.addAll(normalizePersistedLegacySegments(peerLegacy, modern));
        effective.addAll(fromDailySummaries(localSummaries, modern, deviceId, platform));
        return effective;
    }

    public static List<UsageSegment> fromDailySummaries(
            List<DailySummary> summaries,
            List<UsageSegment> existingSegments,
            String deviceId,
            String platform) {
        Set<String> existingLegacySegmentIds = new HashSet<>();
        List<UsageSegment> existingForDevice = new ArrayList<>();
        if (existingSegments != null) {
            for (UsageSegment segment : existingSegments) {
                if (segment != null
                        && safe(deviceId).equals(segment.deviceId)
                        && SOURCE.equals(segment.source)
                        && !safe(segment.segmentId).isEmpty()) {
                    existingLegacySegmentIds.add(segment.segmentId);
                }
                if (segment != null && safe(deviceId).equals(segment.deviceId)) {
                    existingForDevice.add(segment);
                }
            }
        }

        List<UsageSegment> values = new ArrayList<>();
        if (summaries == null) {
            return values;
        }

        for (DailySummary summary : summaries) {
            if (summary == null || summary.date == null || summary.date.trim().isEmpty()) {
                continue;
            }
            for (UsageSegment segment : fromDailySummary(
                    summary,
                    deviceId,
                    platform,
                    coveredModernSecondsByHour(summary.date, existingForDevice))) {
                if (existingLegacySegmentIds.add(segment.segmentId)) {
                    values.add(segment);
                }
            }
        }
        return values;
    }

    private static List<UsageSegment> fromDailySummary(
            DailySummary summary,
            String deviceId,
            String platform,
            long[] coveredSecondsByHour) {
        List<UsageSegment> values = new ArrayList<>();
        long[] hourly = summary.hourlySeconds == null ? new long[24] : summary.hourlySeconds.clone();
        long hourlyTotal = 0L;
        for (long value : hourly) {
            hourlyTotal += Math.max(0L, value);
        }
        if (hourlyTotal <= 0L && summary.totalSeconds > 0L) {
            hourly = distributeAcrossDay(summary.totalSeconds);
        }

        LocalDate date = LocalDate.parse(summary.date);
        for (int hour = 0; hour < 24; hour++) {
            long seconds = Math.max(0L, Math.min(3600L, hourly[hour] - coveredSecondsByHour[hour]));
            if (seconds <= 0L) {
                continue;
            }
            long startUnixSeconds = LocalDateTime.of(date, java.time.LocalTime.of(hour, 0))
                    .atZone(ZoneId.systemDefault())
                    .toEpochSecond();
            long endUnixSeconds = startUnixSeconds + seconds;
            values.add(new UsageSegment(
                    UsageSegmentId.create(deviceId, SOURCE, startUnixSeconds, endUnixSeconds),
                    deviceId,
                    platform == null || platform.trim().isEmpty() ? "android" : platform,
                    SOURCE,
                    startUnixSeconds,
                    endUnixSeconds,
                    summary.date,
                    startUnixSeconds,
                    endUnixSeconds));
        }
        return values;
    }

    private static long[] coveredModernSecondsByHour(String dateValue, List<UsageSegment> existingSegments) {
        long[] covered = new long[24];
        if (dateValue == null || dateValue.trim().isEmpty()) {
            return covered;
        }
        for (UsageSegment segment : existingSegments) {
            if (segment == null
                    || SOURCE.equals(segment.source)
                    || !dateValue.equals(segment.localDate)
                    || segment.endUnixSeconds <= segment.startUnixSeconds) {
                continue;
            }
            addCoveredSeconds(covered, segment);
        }
        for (int hour = 0; hour < covered.length; hour++) {
            covered[hour] = Math.max(0L, Math.min(3600L, covered[hour]));
        }
        return covered;
    }

    private static void addCoveredSeconds(long[] covered, UsageSegment segment) {
        ZonedDateTime start = java.time.Instant.ofEpochSecond(segment.startUnixSeconds).atZone(ZoneId.systemDefault());
        ZonedDateTime end = java.time.Instant.ofEpochSecond(segment.endUnixSeconds).atZone(ZoneId.systemDefault());
        if (!end.isAfter(start)) {
            return;
        }

        ZonedDateTime cursor = start;
        while (cursor.isBefore(end)) {
            ZonedDateTime hourEnd = cursor.withMinute(0).withSecond(0).withNano(0).plusHours(1);
            ZonedDateTime sliceEnd = end.isBefore(hourEnd) ? end : hourEnd;
            covered[cursor.getHour()] += Math.max(0L, java.time.Duration.between(cursor, sliceEnd).getSeconds());
            cursor = sliceEnd;
        }
    }

    private static List<UsageSegment> normalizePersistedLegacySegments(
            List<UsageSegment> legacySegments,
            List<UsageSegment> modernSegments) {
        Map<String, LegacyHourValue> values = new HashMap<>();
        for (UsageSegment segment : legacySegments) {
            if (!isUsableLegacy(segment)) {
                continue;
            }
            int hour = java.time.Instant.ofEpochSecond(segment.startUnixSeconds)
                    .atZone(ZoneId.systemDefault())
                    .getHour();
            String key = segment.deviceId + "|" + segment.platform + "|" + segment.localDate + "|" + hour;
            LegacyHourValue value = values.get(key);
            if (value == null) {
                value = new LegacyHourValue(segment.deviceId, segment.platform, segment.localDate, hour);
                values.put(key, value);
            }
            value.seconds = Math.max(value.seconds, segment.durationSeconds());
        }

        List<UsageSegment> normalized = new ArrayList<>();
        for (LegacyHourValue value : values.values()) {
            long[] covered = coveredModernSecondsByHour(value.date, modernSegmentsForDevice(modernSegments, value.deviceId));
            long seconds = Math.max(0L, Math.min(3600L, value.seconds - covered[value.hour]));
            if (seconds <= 0L) {
                continue;
            }
            LocalDate date = LocalDate.parse(value.date);
            long startUnixSeconds = LocalDateTime.of(date, java.time.LocalTime.of(value.hour, 0))
                    .atZone(ZoneId.systemDefault())
                    .toEpochSecond();
            long endUnixSeconds = startUnixSeconds + seconds;
            normalized.add(new UsageSegment(
                    UsageSegmentId.create(value.deviceId, SOURCE, startUnixSeconds, endUnixSeconds),
                    value.deviceId,
                    value.platform == null || value.platform.trim().isEmpty() ? "android" : value.platform,
                    SOURCE,
                    startUnixSeconds,
                    endUnixSeconds,
                    value.date,
                    startUnixSeconds,
                    endUnixSeconds));
        }
        return normalized;
    }

    private static List<UsageSegment> modernSegmentsForDevice(List<UsageSegment> segments, String deviceId) {
        List<UsageSegment> values = new ArrayList<>();
        for (UsageSegment segment : segments) {
            if (segment != null && safe(deviceId).equals(segment.deviceId)) {
                values.add(segment);
            }
        }
        return values;
    }

    private static boolean isUsableLegacy(UsageSegment segment) {
        return segment != null
                && !safe(segment.deviceId).isEmpty()
                && SOURCE.equals(segment.source)
                && !safe(segment.localDate).isEmpty()
                && segment.endUnixSeconds > segment.startUnixSeconds;
    }

    private static UsageSegment cloneSegment(UsageSegment segment) {
        return new UsageSegment(
                segment.segmentId,
                segment.deviceId,
                segment.platform,
                segment.source,
                segment.startUnixSeconds,
                segment.endUnixSeconds,
                segment.localDate,
                segment.createdAtUnixSeconds,
                segment.updatedAtUnixSeconds);
    }

    private static final class LegacyHourValue {
        final String deviceId;
        final String platform;
        final String date;
        final int hour;
        long seconds;

        LegacyHourValue(String deviceId, String platform, String date, int hour) {
            this.deviceId = deviceId;
            this.platform = platform;
            this.date = date;
            this.hour = hour;
        }
    }

    private static long[] distributeAcrossDay(long totalSeconds) {
        long[] hourly = new long[24];
        long remaining = Math.max(0L, totalSeconds);
        for (int hour = 0; hour < hourly.length && remaining > 0L; hour++) {
            hourly[hour] = Math.min(3600L, remaining);
            remaining -= hourly[hour];
        }
        return hourly;
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }
}

File: src/EyeTimeTracker.Core/Sync/LegacyUsageSegments.cs
using EyeTimeTracker.Core.Models;

namespace EyeTimeTracker.Core.Sync;

public static class LegacyUsageSegments
{
    public const string Source = "legacy-summary";

    public static List<UsageSegment> NormalizeEffectiveSegments(
        IEnumerable<UsageSegment> storedSegments,
        IEnumerable<DailyRecord> localRecords,
        string deviceId,
        string platform)
    {
        var stored = storedSegments?.ToList() ?? new List<UsageSegment>();
        var modernSegments = stored
            .Where(segment => !string.Equals(segment.Source, Source, StringComparison.Ordinal))
            .Select(Clone)
            .ToList();
        var peerLegacySegments = stored
            .Where(segment => string.Equals(segment.Source, Source, StringComparison.Ordinal))
            .Where(segment => !string.Equals(segment.DeviceId, deviceId, StringComparison.Ordinal))
            .ToList();

        var effective = new List<UsageSegment>();
        effective.AddRange(modernSegments);
        effective.AddRange(NormalizePersistedLegacySegments(peerLegacySegments, modernSegments));
        effective.AddRange(FromDailyRecords(
            localRecords,
            modernSegments.Where(segment => string.Equals(segment.DeviceId, deviceId, StringComparison.Ordinal)),
            deviceId,
            platform));
        return effective;
    }

    public static List<UsageSegment> FromDailyRecords(
        IEnumerable<DailyRecord> records,
        IEnumerable<UsageSegment> existingSegments,
        string deviceId,
        string platform)
    {
        var existingLegacySegmentIds = existingSegments
            .Where(segment => string.Equals(segment.DeviceId, deviceId, StringComparison.Ordinal))
            .Where(segment => string.Equals(segment.Source, Source, StringComparison.Ordinal))
            .Where(segment => !string.IsNullOrWhiteSpace(segment.SegmentId))
            .Select(segment => segment.SegmentId)
            .ToHashSet(StringComparer.Ordinal);

        var existingForDevice = existingSegments
            .Where(segment => string.Equals(segment.DeviceId, deviceId, StringComparison.Ordinal))
            .ToList();

        return records
            .Where(record => record.Date != default)
            .SelectMany(record => FromDailyRecord(
                record,
                deviceId,
                platform,
                CoveredModernSecondsByHour(record.Date, existingForDevice)))
            .Where(segment => existingLegacySegmentIds.Add(segment.SegmentId))
            .ToList();
    }

    private static IEnumerable<UsageSegment> FromDailyRecord(
        DailyRecord record,
        string deviceId,
        string platform,
        long[] coveredSecondsByHour)
    {
        var hourly = NormalizeHourly(record.HourlySeconds);
        var hourlyTotal = hourly.Sum();
        if (hourlyTotal <= 0 && record.TotalSeconds > 0)
        {
            hourly = DistributeAcrossDay(record.TotalSeconds);
        }

        for (var hour = 0; hour < hourly.Length; hour++)
        {
            var seconds = Math.Clamp(hourly[hour] - coveredSecondsByHour[hour], 0, 3600);
            if (seconds <= 0)
            {
                continue;
            }

            var localStart = record.Date.ToDateTime(TimeOnly.FromTimeSpan(TimeSpan.FromHours(hour)));
            var start = new DateTimeOffset(localStart, TimeZoneInfo.Local.GetUtcOffset(localStart));
            var end = start.AddSeconds(seconds);
            yield return new UsageSegment
            {
                SegmentId = UsageSegmentId.Create(deviceId, Source, start, end),
                DeviceId = deviceId,
                Platform = string.IsNullOrWhiteSpace(platform) ? "windows" : platform,
                Source = Source,
                StartUnixSeconds = start.ToUnixTimeSeconds(),
                EndUnixSeconds = end.ToUnixTimeSeconds(),
                LocalDate = record.Date,
                CreatedAtUnixSeconds = start.ToUnixTimeSeconds(),
                UpdatedAtUnixSeconds = end.ToUnixTimeSeconds()
            };
        }
    }

    private static long[] CoveredModernSecondsByHour(DateOnly date, IEnumerable<UsageSegment> existingSegments)
    {
        var covered = new long[24];
        foreach (var segment in existingSegments)
        {
            if (segment.LocalDate != date
                || string.Equals(segment.Source, Source, StringComparison.Ordinal)
                || segment.EndUnixSeconds <= segment.StartUnixSeconds)
            {
                continue;
            }

            AddCoveredSeconds(covered, segment);
        }

        for (var hour = 0; hour < covered.Length; hour++)
        {
            covered[hour] = Math.Clamp(covered[hour], 0, 3600);
        }

        return covered;
    }

    private static void AddCoveredSeconds(long[] covered, UsageSegment segment)
    {
        var start = DateTimeOffset.FromUnixTimeSeconds(segment.StartUnixSeconds).ToLocalTime();
        var end = DateTimeOffset.FromUnixTimeSeconds(segment.EndUnixSeconds).ToLocalTime();
        if (end <= start)
        {
            return;
        }

        var cursor = start;
        while (cursor < end)
        {
            var hourStart = new DateTimeOffset(
                cursor.Year,
                cursor.Month,
                cursor.Day,
                cursor.Hour,
                0,
                0,
                cursor.Offset);
            var hourEnd = hourStart.AddHours(1);
            var sliceEnd = end < hourEnd ? end : hourEnd;
            covered[cursor.Hour] += Math.Max(0, (long)(sliceEnd - cursor).TotalSeconds);
            cursor = sliceEnd;
        }
    }

    private static long[] NormalizeHourly(long[]? source)
    {
        var hourly = new long[24];
        if (source is not null)
        {
            Array.Copy(source, hourly, Math.Min(hourly.Length, source.Length));
        }

        return hourly;
    }

    private static long[] DistributeAcrossDay(long totalSeconds)
    {
        var hourly = new long[24];
        var remaining = Math.Max(0, totalSeconds);
        for (var hour = 0; hour < hourly.Length && remaining > 0; hour++)
        {
            hourly[hour] = Math.Min(3600, remaining);
            remaining -= hourly[hour];
        }

        return hourly;
    }

    private static IEnumerable<UsageSegment> NormalizePersistedLegacySegments(
        IEnumerable<UsageSegment> legacySegments,
        IEnumerable<UsageSegment> modernSegments)
    {
        var modern = modernSegments.ToList();
        return legacySegments
            .Where(IsUsableLegacy)
            .GroupBy(segment => new LegacyHourKey(
                segment.DeviceId,
                segment.Platform,
                segment.LocalDate,
                DateTimeOffset.FromUnixTimeSeconds(segment.StartUnixSeconds).ToLocalTime().Hour))
            .Select(group => CreateNormalizedLegacySegment(group.Key, group.Max(segment => segment.DurationSeconds), modern))
            .Where(segment => segment is not null)
            .Select(segment => segment!);
    }

    private static UsageSegment? CreateNormalizedLegacySegment(
        LegacyHourKey key,
        long legacySeconds,
        IEnumerable<UsageSegment> modernSegments)
    {
        var covered = CoveredModernSecondsByHour(
            key.Date,
            modernSegments.Where(segment => string.Equals(segment.DeviceId, key.DeviceId, StringComparison.Ordinal)))[key.Hour];
        var seconds = Math.Clamp(legacySeconds - covered, 0, 3600);
        if (seconds <= 0)
        {
            return null;
        }

        var localStart = key.Date.ToDateTime(TimeOnly.FromTimeSpan(TimeSpan.FromHours(key.Hour)));
        var start = new DateTimeOffset(localStart, TimeZoneInfo.Local.GetUtcOffset(localStart));
        var end = start.AddSeconds(seconds);
        return new UsageSegment
        {
            SegmentId = UsageSegmentId.Create(key.DeviceId, Source, start, end),
            DeviceId = key.DeviceId,
            Platform = string.IsNullOrWhiteSpace(key.Platform) ? "windows" : key.Platform,
            Source = Source,
            StartUnixSeconds = start.ToUnixTimeSeconds(),
            EndUnixSeconds = end.ToUnixTimeSeconds(),
            LocalDate = key.Date,
            CreatedAtUnixSeconds = start.ToUnixTimeSeconds(),
            UpdatedAtUnixSeconds = end.ToUnixTimeSeconds()
        };
    }

    private static bool IsUsableLegacy(UsageSegment segment)
    {
        return !string.IsNullOrWhiteSpace(segment.DeviceId)
            && string.Equals(segment.Source, Source, StringComparison.Ordinal)
            && segment.LocalDate != default
            && segment.EndUnixSeconds > segment.StartUnixSeconds;
    }

    private static UsageSegment Clone(UsageSegment segment)
    {
        return new UsageSegment
        {
            SegmentId = segment.SegmentId,
            DeviceId = segment.DeviceId,
            Platform = segment.Platform,
            Source = segment.Source,
            StartUnixSeconds = segment.StartUnixSeconds,
            EndUnixSeconds = segment.EndUnixSeconds,
            LocalDate = segment.LocalDate,
            CreatedAtUnixSeconds = segment.CreatedAtUnixSeconds,
            UpdatedAtUnixSeconds = segment.UpdatedAtUnixSeconds
        };
    }

    private sealed record LegacyHourKey(string DeviceId, string Platform, DateOnly Date, int Hour);
}
